// nps/dist/bin-utils/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadConfig = exports.preloadModule = exports.getModuleRequirePath = exports.help = exports.initialize = undefined;

var _path = require('path');

var _fs = require('fs');

var _lodash = require('lodash');

var _typeDetect = require('type-detect');

var _typeDetect2 = _interopRequireDefault(_typeDetect);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _jsYaml = require('js-yaml');

var _commonTags = require('common-tags');

var _getLogger = require('../get-logger');

var _getLogger2 = _interopRequireDefault(_getLogger);

var _resolveScriptObjectToString = require('../resolve-script-object-to-string');

var _initialize = require('./initialize');

var _initialize2 = _interopRequireDefault(_initialize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var log = (0, _getLogger2.default)();

/**
 * Attempts to load the given module. This is used for the
 * --require functionality of the CLI
 * @param  {String} moduleName The module to attempt to require
 * @return {*} The required module
 */
var preloadModule = getAttemptModuleRequireFn(function (moduleName, requirePath) {
  log.warn({
    message: _chalk2.default.yellow(_commonTags.oneLine`
        Unable to preload "${moduleName}".
        Attempted to require as "${requirePath}"
      `),
    ref: 'unable-to-preload-module'
  });
  return undefined;
});

var loadJSConfig = getAttemptModuleRequireFn(function onFail(configPath, requirePath) {
  log.error({
    message: _chalk2.default.red(_commonTags.oneLine`
        Unable to find JS config at "${configPath}".
        Attempted to require as "${requirePath}"
      `),
    ref: 'unable-to-find-config'
  });
  return undefined;
});

/**
 * Attempts to load the config and logs an error if there's a problem
 * @param {String} configPath The path to attempt to require the config from
 * @param {*} input the input to pass to the config if it's a function
 * @return {Object} The config
 */
function loadConfig(configPath, input) {
  if (configPath.endsWith('.yml')) {
    return loadYAMLConfig(configPath);
  }

  var config = loadJSConfig(configPath);
  if ((0, _lodash.isUndefined)(config)) {
    // let the caller deal with this
    return config;
  }
  var typeMessage = void 0;
  if ((0, _lodash.isFunction)(config)) {
    config = config(input);
    typeMessage = _commonTags.oneLine`
      Your config was a function that returned
      a data type of "${(0, _typeDetect2.default)(config)}"
    `;
  } else {
    typeMessage = `Your config data type was "${(0, _typeDetect2.default)(config)}"`;
  }
  if (!(0, _lodash.isPlainObject)(config)) {
    log.error({
      message: _chalk2.default.red(_commonTags.oneLine`
          The package-scripts configuration
          ("${configPath}") must be an object
          or a function that returns an object.
        `),
      ref: 'config-must-be-an-object'
    });
    throw new Error(typeMessage);
  }
  return config;
}

exports.initialize = _initialize2.default;
exports.help = help;
exports.getModuleRequirePath = getModuleRequirePath;
exports.preloadModule = preloadModule;
exports.loadConfig = loadConfig;

/****** implementations ******/

function loadYAMLConfig(configPath) {
  try {
    return (0, _jsYaml.safeLoad)((0, _fs.readFileSync)(configPath, 'utf8'));
  } catch (e) {
    if (e.constructor.name === 'YAMLException') {
      throw e;
    }
    log.error({
      message: _chalk2.default.red(`Unable to find YML config at "${configPath}".`),
      ref: 'unable-to-find-config'
    });
    return undefined;
  }
}

/**
 * Determines the proper require path for a module.
 * If the path starts with `.` then it is resolved with process.cwd()
 * @param  {String} moduleName The module path
 * @return {String} the module path to require
 */
function getModuleRequirePath(moduleName) {
  return moduleName[0] === '.' ? (0, _path.resolve)(process.cwd(), moduleName) : moduleName;
}

function getAttemptModuleRequireFn(onFail) {
  return function attemptModuleRequire(moduleName) {
    var requirePath = getModuleRequirePath(moduleName);
    try {
      return requireDefaultFromModule(requirePath);
    } catch (e) {
      if (e.constructor.name === 'SyntaxError') {
        throw e;
      }
      return onFail(moduleName, requirePath);
    }
  };
}

/**
 * Requires the given module and returns the `default` if it's an `__esModule`
 * @param  {String} modulePath The module to require
 * @return {*} The required module (or it's `default` if it's an `__esModule`)
 */
function requireDefaultFromModule(modulePath) {
  /* eslint global-require:0,import/no-dynamic-require:0 */
  var mod = require(modulePath);
  if (mod.__esModule) {
    return mod.default;
  } else {
    return mod;
  }
}

function help(_ref) {
  var scripts = _ref.scripts;

  var availableScripts = getAvailableScripts(scripts);
  var filteredScripts = availableScripts.filter(function (script) {
    return !script.hiddenFromHelp;
  });
  var scriptLines = filteredScripts.map(function (_ref2) {
    var name = _ref2.name,
        description = _ref2.description,
        script = _ref2.script;

    var coloredName = _chalk2.default.green(name);
    var coloredScript = _chalk2.default.gray(script);
    var line = void 0;
    if (description) {
      line = [coloredName, _chalk2.default.white(description), coloredScript];
    } else {
      line = [coloredName, coloredScript];
    }
    return line.join(' - ').trim();
  });
  if (scriptLines.length) {
    var topMessage = 'Available scripts (camel or kebab case accepted)';
    var message = `${topMessage}\n\n${scriptLines.join('\n')}`;
    return message;
  } else {
    return _chalk2.default.yellow('There are no scripts available');
  }
}

function getAvailableScripts(config) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];

  var excluded = ['description', 'script', 'default'];
  return Object.keys(config).reduce(function (scripts, key) {
    var val = config[key];
    if ((0, _lodash.includes)(excluded, key)) {
      return scripts;
    }
    var scriptObj = (0, _resolveScriptObjectToString.resolveScriptObjectToScript)(val);
    var prefixed = [].concat(_toConsumableArray(prefix), [key]);
    if (scriptObj) {
      var description = scriptObj.description,
          script = scriptObj.script,
          _scriptObj$hiddenFrom = scriptObj.hiddenFromHelp,
          hiddenFromHelp = _scriptObj$hiddenFrom === undefined ? false : _scriptObj$hiddenFrom;

      scripts = [].concat(_toConsumableArray(scripts), [{ name: prefixed.join('.'), description, script, hiddenFromHelp }]);
    }
    if ((0, _lodash.isPlainObject)(val)) {
      return [].concat(_toConsumableArray(scripts), _toConsumableArray(getAvailableScripts(val, prefixed)));
    }
    return scripts;
  }, []);
}
